### ISSUE TITLE
question about warning on page 290

### ISSUE THREAD

**Issue by eliot1785** [amueller/introduction_to_ml_with_python#49]
The book says: "For simplicity, we changed the threshold value based on test set
results in the code above. In practice, you need to use a hold-out
validation set, not the test set. As with any other parameter, setting
a decision threshold on the test set is likely to yield overly optimistic
results. Use a validation set or cross-validation instead."

I'm a little confused by this. Are you really saying we should experiment with different threshold values using our holdout set? I thought that the holdout was supposed to be set aside until you're completely done tweaking hyperparameters (of which threshold is one, if I'm not mistaken), and then you run it once on the holdout and that's it. 

Or to put it another way, isn't the test set actually where we would want to experiment with parameters? 

Also a little confused by the final sentence. There's lots of different terminology, but my impression is you either use a train/test split, or you use something more sophisticated like cross-validation.  Setting the threshold in CV would be equivalent to setting the threshold in test. I don't see why CV would necessarily be a "better place" for avoiding overfitting, as this implies.

Then again, I'm a noob so I could be missing something :)

**Comment by amueller** [comment]
I think this is more about the different terminology of hold-out vs test-set. What I call "test set" in the book is the final set that you only look at once. The validation set is a separate split that can be used for tuning. Instead of the validation set, you can also use cross-validation.

**Comment by eliot1785** [comment]
So when you say "In practice, you need to use a hold-out validation set, not the test set," the "hold-out validation set" is less 'final' than the 'test set'?

**Comment by amueller** [comment]
yes. I think it's explained more in chapter 5 and figure 5-5.
